README.md
# PCEAngelShooter

# This is my first proper project on PCE 
It is just a game I made to have some fun :) so basically I will put everything I'm learning and hopefully it will be playable ;)
The sprites aren't mine they just act as placeholders


mednafen/ChangeLog
0.8.D:
	SMS:  Some state that wasn't being saved in save states now is, which should fix netplay desynch problems with SMS/GG games.

        Fixed a couple of remotely-exploitable(if connected to a malicious server) stack manipulation bugs in the network play code.

	Fixed an incorrect object creation bug in pce/adpcm.cpp.

	NES:  Added a missing CPU emulator variable to save states.

	PCE:  Added a missing CPU emulator variable to save states.

0.8.D-rc1:
	Lynx:  Fixed a bug in the cart loader code that would cause a crash if the ROM bank size was larger than the actual data available in the
	file(as is the case with some homebrew programs).  Thanks to "Wookie" for the patch.

	Build files were regenerated using autoconf 2.64 and aclocal 1.11(previously, they were generated with autoconf 2.61 and aclocal 1.10.1).

	Fixed a crashing problem when entering an invalid menu choice("0") in the cheat interface.  Thanks to
	tsenart for reporting the bug.

	GB:  The GameBoy module now respects the "filesys.disablesavegz" setting in respect to saved
	battery-backed RAM.

	Added support for "lurkers" on the network play server.  Previous versions of Mednafen don't lack support for this per se, but there
	would be cosmetic issues with status messages printed to the internal console.

	SexyAL:  Fixed a bug affecting the return value from RawCanWrite() in the ALSA driver.  The returned value was typically too
		 small by a factor of 4.  The effects of this bug included potential long periods of garbled sound
		 during netplay.

		 Fixed the return value from RawCanWrite() in the JACK driver.  It was being clamped to a value
		 that was too small by a factor of 4; however, the clamp value was already excessively large in a way
		 that this bug would should have only been triggered if the "soundbufsize" setting was excessively large.
		 The effects of this bug would be similar to the ALSA RawCanWrite() bug.

		 The ALSA and OSS drivers will now try to set audio output to 2 channels if the source data only has 1 channel, and 16-bit signed if the
		 source data is 8-bit(automatic conversion is done).  This is done to allow for lower period/fragment sizes, as, in ALSA's internals at least,
		 the minimum period sizes are expressed in bytes, not sound frames.

		 The ALSA and OSS drivers will now try to set lower period/fragment sizes than previous versions of Mednafen did.  With default settings, for
		 ALSA, the new period/fragment size is 50% of what it was before, and for OSS, 25%.  Also, there's a new setting to override
		 the SexyAL's driver's preferred period/fragment sizes, named "sound.period_time"(default value of 0: no override). 
		 The period/fragment size is expressed in microseconds.  If the new, lower fragment sizes cause problems, the setting can be changed to "2666"
		 to approximate the fragment size selection in previous versions of Mednafen when using ALSA output, and "5333" when using OSS output.

		 Added a workaround to the OSS driver for a bug in ALSA(and hence, ALSA's in-kernel OSS emulation) that could cause the emulator to run far
		 too fast for a short period of time if a buffer underflow occurred.

		 The ALSA's driver's RawCanWrite() method now(finally) uses snd_pcm_avail_update() instead of snd_pcm_delay().
		 This should improve performance and frameskipping behavior when the ALSA output is not routed directly to a physical device, such as the case with
		 PulseAudio(though PulseAudio is still not recommended :b).

0.8.C:

	SMS:  Fixed the input and reset emulation when the territory was
	set to "domestic"(Japan) and FM emulation was enabled; the emulated system was erroneously
	being treated as a Game Gear in a few places when those conditions were met.

	GB:  Fixed initialization of "HRAM" on virtual power toggle/reset.

	GB:  Fixed a bug(still many left) in the external RAM handling code, fixing an
	emulator crash with "Wizards & Warriors - Chapter X".

0.8.C-rc1:

	NGP: Fixed an unaligned memory access bug in the memory r/w handlers, causing crashes on platforms
	that do not support unaligned memory accesses.

	NES: Fixed an intialization bug in mapper 193.

	Added define "ALWAYS_INLINE" to force the compiler to always inline speed-critical functions, and added it to
	many emulation functions in that category.  gcc can be a pain in the butt.

	GBA:  Fixed a bug that would cause data corruption and the loaded game to crash whenever a state was
	saved(but the state itself, and loading it, was fine) on a big-endian platform.

	Fixed several instances of mismatches between format strings and passed argument types, which could cause
	nonsensical information to be printed on big-endian platforms, and perhaps some little-endian platforms as well.

	Fixed the minimum field-width specifier on the statements that print the loaded game's CRC32 in several
	places.

	Added the "format(...)" function attribute to the MDFN_DispMessage(), MDFN_PrintError(), and MDFN_printf()
	function prototypes to catch format string errors.

	Fixed the order of the sound drivers in the documentation(it was erroneously implying that OSS had a higher
	priority than ALSA for the default sound driver), and added various notes to explain the sound drivers.

	If an invalid "sounddriver" is selected, Mednafen will now print an error message and not initialize sound
	instead of silently selecting the default sound driver.

	SexyAL: Altered the argument order of SexiALI_Convert() to make more sense.

		Fixed the broken conversion code used when the frame count passed to Write() is greater than 2048.

		Fixed a bug/design flaw in the SexyAL SDL sound driver that can cause a crash during program
		deinitialization.

		Added support for "float" output format to the ALSA driver.

		The ALSA driver will now try all formats that SexyAL's conversion code supports if the desired
		format is not supported by the output device, rather that simply erroring out as before.


include/pce/AC.h
#ifndef _AC_H
#define _AC_H

/*
 * Arcade Card fastcall defines
   Arcade Card inline defines
 */


/*
 *	Unused test. single 24bit argument passing.
 */
#pragma fastcall test( dword cl:ax|bx  ) nop;


/*
 *	Arcade card address reg function: 24bit value, 1 byte(high) and 1 word(mid/low). 
 */ 

void __fastcall __nop ac_addr_reg0( unsigned char high<ac_reg_1_high>, unsigned int low<ac_reg_1_low> );
void __fastcall __nop ac_addr_reg1( unsigned char high<ac_reg_2_high>, unsigned int low<ac_reg_2_low> );
void __fastcall __nop ac_addr_reg2( unsigned char high<ac_reg_3_high>, unsigned int low<ac_reg_3_low> );
void __fastcall __nop ac_addr_reg3( unsigned char high<ac_reg_4_high>, unsigned int low<ac_reg_4_low> );


/*
 *	Arcade card full reg function: address byte high, address word mid/low, word offset,
 *	word increment, byte control. *Inlined*
 *
 */
void __fastcall __nop ac_full_reg0( unsigned char high<ac_reg_1_high>, unsigned int low<ac_reg_1_low>, unsigned int offset_l<ac_reg_1_offset_l>, unsigned int incrmt_l<ac_reg_1_incrmt_l>, unsigned char control<ac_reg_1_control>);
void __fastcall __nop ac_full_reg1( unsigned char high<ac_reg_2_high>, unsigned int low<ac_reg_2_low>, unsigned int offset_l<ac_reg_2_offset_l>, unsigned int incrmt_l<ac_reg_2_incrmt_l>, unsigned char control<ac_reg_2_control>);
void __fastcall __nop ac_full_reg2( unsigned char high<ac_reg_3_high>, unsigned int low<ac_reg_3_low>, unsigned int offset_l<ac_reg_3_offset_l>, unsigned int incrmt_l<ac_reg_3_incrmt_l>, unsigned char control<ac_reg_3_control>);
void __fastcall __nop ac_full_reg3( unsigned char high<ac_reg_4_high>, unsigned int low<ac_reg_4_low>, unsigned int offset_l<ac_reg_4_offset_l>, unsigned int incrmt_l<ac_reg_4_incrmt_l>, unsigned char control<ac_reg_4_control>);


/*
 *	Arcade card increment reg function: word size. *Inlined*
 */
void __fastcall __nop ac_inc_reg0( unsigned int incrmt_l<ac_reg_1_incrmt_l> );
void __fastcall __nop ac_inc_reg1( unsigned int incrmt_l<ac_reg_2_incrmt_l> );
void __fastcall __nop ac_inc_reg2( unsigned int incrmt_l<ac_reg_3_incrmt_l> );
void __fastcall __nop ac_inc_reg3( unsigned int incrmt_l<ac_reg_4_incrmt_l> );


/*
 *	Arcade card offset reg function: word size. *Inlined*
 */
void __fastcall __nop ac_offset_reg0( unsigned int offset_l<ac_reg_1_offset_l> );
void __fastcall __nop ac_offset_reg1( unsigned int offset_l<ac_reg_2_offset_l> );
void __fastcall __nop ac_offset_reg2( unsigned int offset_l<ac_reg_3_offset_l> );
void __fastcall __nop ac_offset_reg3( unsigned int offset_l<ac_reg_4_offset_l> );


/*
 *	Arcade card control reg function: byte size. *Inlined*
 */
void __fastcall __nop ac_control_reg0( unsigned char control<ac_reg_1_control_l> );
void __fastcall __nop ac_control_reg1( unsigned char control<ac_reg_2_control_l> );
void __fastcall __nop ac_control_reg2( unsigned char control<ac_reg_3_control_l> );
void __fastcall __nop ac_control_reg3( unsigned char control<ac_reg_4_control_l> );


/*
 *	CD sector to Arcade card transfer.
 */
void __fastcall ac_cd_xfer( unsigned char ac_reg<bl>, unsigned char sector_addr_h<cl>, unsigned int sector_addr_l<dx>, unsigned char sectors<al> );

void __fastcall ac_vram_xfer(unsigned char ac_reg<al>, unsigned int vram_addr<bx>, unsigned int num_bytes<cx>, unsigned char size<dl> );
void __fastcall ac_vram_xfer(unsigned char ac_reg<al>, unsigned int vram_addr<bx>, unsigned int num_bytes<cx>, unsigned char size<dl>, unsigned char flags<ah> );

void __fastcall ac_vram_dma( unsigned char ac_reg<al>, unsigned int vram_addr<bx>, unsigned int num_bytes<cx> );
void __fastcall ac_vram_dma( unsigned char ac_reg<al>, unsigned int vram_addr<bx>, unsigned int num_bytes<cx>, unsigned char flags<ah> );


void __fastcall ac_vram_copy( unsigned char ac_reg<al>, unsigned int vram_addr<bx>, unsigned int bytes<cx> );
void __fastcall ac_vram_copy( unsigned char ac_reg<al>, unsigned int vram_addr<bx>, unsigned int bytes<cx>, unsigned char flags<ah> );


void __fastcall ac_vce_copy( unsigned char ac_reg<al>, unsigned int start_color<bx>, unsigned int num_colors<cx> );


/*
 * Arcade Card defines
 */
#define AC_REG0	0x00
#define AC_REG1	0x01
#define AC_REG2	0x02
#define AC_REG3	0x03

/*
 * Block Transfer defines for use with Arcade Call routines
 */
#define TIA_OP			0xE300
#define TAI_OP			0xF300
#define TIN_OP			0xD300
#define TII_OP			0x7300
#define TDD_OP			0xC300
#define RTS			0x0060

#endif /* _AC_H */


include/pce/sgx.h
#ifndef _SGX_H
#define _SGX_H

/*
 * SGX fastcall defines
 */


void __fastcall sgx_vreg( unsigned char reg<acc> );
void __fastcall sgx_vreg( unsigned char reg<al>, unsigned int data<cx> );

void __fastcall sgx_read_vram( unsigned int vram_offset<ax> );

void __fastcall sgx_load_bat(unsigned int vaddr<di>, int far *bat_data<bl:si>, unsigned char w<cl>, unsigned char h<ch>);

void __fastcall sgx_set_screen_size( unsigned char size<acc> );

void __fastcall sgx_load_vram(unsigned int vaddr <di>, int far *data<bl:si>, int nb<cx>);
 
void __fastcall sgx_set_tile_data(char *tile_ex<di>);
void __fastcall sgx_set_tile_data(char far *tile<bl:si>, int nb_tile<cx>, char far *ptable<al:dx>);

 
void __fastcall sgx_set_map_data(int *ptr<acc>);
void __fastcall sgx_set_map_data(char far *map<bl:si>, int w<ax>, int h<acc>);
void __fastcall sgx_set_map_data(char far *map<bl:si>, int w<ax>, int h<dx>, char wrap<acc>);

void __fastcall sgx_load_map(char x<al>, char y<ah>, int mx<di>, int my<bx>, char w<dl>, char h<dh>);

void __fastcall sgx_scroll(int x<ax>, int y<bx>);

void __fastcall sgx_spr_set( char num<acc> );

void sgx_satb_update(void);
void __fastcall sgx_satb_update( unsigned char max<acc> );

void __fastcall sgx_spr_hide( char num<acc> );

void __fastcall sgx_spr_show( char num<acc> );

void __fastcall sgx_spr_ctrl(char mask<al>, char value<acc>);

void __fastcall vpc_win_size(char window_num<al>, int size<bx>);

void __fastcall vpc_win_reg(char window_num<al>, char var<bl>);


/*
 * SGX defines
 */
#define SGX				0x01
#define VPC_WIN_A			0x00
#define VPC_WIN_B			0x01
#define	VPC_WIN_AB			0x02
#define	VPC_WIN_NONE			0x03
#define	VPC_NORM			0x00
#define	VPC_SPR				0x04
#define	VPC_INV_SPR			0x08
#define VDC1_ON				0x01
#define	VDC1_OFF			0x00
#define VDC2_ON				0x02
#define	VDC2_OFF			0x00
#define VDC_ON				0x03
#define	VDC_OFF				0x00

#endif /* _SGX_H */


include/pce/huc.h
/*
 * backup ram defines
 */

#define  BM_OK             0
#define  BM_NOT_FOUND      1
#define  BM_BAD_CHECKSUM   2
#define  BM_DIR_CORRUPTED  3
#define  BM_FILE_EMPTY     4
#define  BM_FULL           5
#define  BM_NOT_FORMATED   0xFF

#define  BRAM_STARTPTR     0x8010

/*
 * sprite defines
 */

#define  FLIP_X_MASK 0x08
#define  FLIP_Y_MASK 0x80
#define  FLIP_MAS   0x88
#define  SIZE_MAS   0x31

#define  NO_FLIP    0x00
#define  NO_FLIP_X  0x00
#define  NO_FLIP_Y  0x00
#define  FLIP_X     0x08
#define  FLIP_Y     0x80
#define  SZ_16x16   0x00
#define  SZ_16x32   0x10
#define  SZ_16x64   0x30
#define  SZ_32x16   0x01
#define  SZ_32x32   0x11
#define  SZ_32x64   0x31

/*
 * joypad defines
 */

#define	JOY_A		0x01
#define	JOY_I		0x01
#define	JOY_B		0x02
#define	JOY_II		0x02
#define	JOY_SLCT	0x04
#define	JOY_SEL		0x04
#define	JOY_STRT	0x08
#define	JOY_RUN		0x08
#define	JOY_UP		0x10
#define	JOY_RGHT	0x20
#define	JOY_RIGHT	0x20
#define	JOY_DOWN	0x40
#define	JOY_LEFT	0x80

#define	JOY_C		0x0100
#define	JOY_III		0x0100
#define	JOY_D		0x0200
#define	JOY_IV		0x0200
#define	JOY_E		0x0400
#define	JOY_V		0x0400
#define	JOY_F		0x0800
#define	JOY_VI		0x0800

#define JOY_SIXBUT	0x5000
#define JOY_TYPE6	0x5000


/*
 * screen defines
 */
#define	SCR_SIZE_32x32	0
#define	SCR_SIZE_64x32	1
#define	SCR_SIZE_128x32	3
#define	SCR_SIZE_32x64	4
#define	SCR_SIZE_64x64	5
#define	SCR_SIZE_128x64	7

#define	XRES_SHARP	0
#define	XRES_SOFT	4

/*
 * CD defines
 */
#define	CDPLAY_MUTE		0
#define	CDPLAY_REPEAT		1
#define	CDPLAY_NORMAL		2
#define	CDPLAY_ENDOFDISC	0

#define	CDFADE_CANCEL	0
#define	CDFADE_PCM6	8
#define	CDFADE_ADPCM6	10
#define	CDFADE_PCM2	12
#define	CDFADE_ADPCM2	14

#define CDTRK_AUDIO	0
#define CDTRK_DATA	4

/*
 * ADPCM defines
 */
#define	ADPLAY_AUTOSTOP		0
#define	ADPLAY_REPEAT		0x80

#define ADPLAY_FREQ_16KHZ	0xE
#define ADPLAY_FREQ_10KHZ	0xD
#define ADPLAY_FREQ_8KHZ	0xC
#define ADPLAY_FREQ_6KHZ	0xB
#define ADPLAY_FREQ_5KHZ	0xA

#define ADREAD_RAM	0
#define ADREAD_VRAM	0xFF

#define ADWRITE_RAM	0
#define ADWRITE_VRAM	0xFF

void __fastcall set_map_pals(char far *pal<bl:si>);
void set_map_tile_type(unsigned char);
void set_map_tile_base(unsigned int vaddr);

#define IRQ_VSYNC	8
#define IRQ_HSYNC	16
#define IRQ_TIMER	4

unsigned char mem_mapdatabank(unsigned char new_bank);
unsigned int mem_mapdatabanks(unsigned int new_banks);

void timer_set(unsigned char cnt);
void timer_start(void);
void timer_stop(void);
unsigned char timer_get(void);

void irq_enable_user(unsigned char irq);
void irq_disable_user(unsigned char irq);

void irq_enable(unsigned char irq);
void irq_disable(unsigned char irq);

